### PR TITLE
Cursor server association support

### PR DESCRIPTION
**Pull request by tolsen**
This adds support to mongonet for helping clients to associate a cursor with a particular server.

Louisa, I am assigning to you for now because Tomer is out until Sunday.  If we don't need this merged until next week then we can add Tomer as a reviewer.

### CODE REVIEW

File: wire.go [hunk 1]
```diff
@@ -145,5 +156,16 @@ type MessageMessage struct {
...
+func (mm *MessageMessage) BodyDoc() bsoncore.Document {
+	// when parsed in in parseMessageMessage we checked that there was exactly one BodySection
+	// as such we can stop as soon as we find it
+	for _, sec := range mm.Sections {
+		if bodySection, ok := sec.(*BodySection); ok && bodySection != nil {
+			return bsoncore.Document(bodySection.Body.BSON)
+		}
+	}
+
+	panic(fmt.Errorf("unreachable"))
```
Comment (TomerYakir) [inline wire.go:170]: could we rephrase as something like `no body section found for OP_MSG`?

File: session.go [hunk 2]
```diff
@@ -174,10 +172,12 @@ func (s *Session) RespondToCommand(clientMessage Message, doc SimpleBSON) error
 					doc,
 				},
 			},
-			bsoncore.Document(doc.BSON),
 		}
 		return SendMessage(rm, s.conn)
 
+	case OP_GET_MORE:
+		return errors.New("Internal error")
```
Comment (TomerYakir) [inline session.go:179]: [q] do we want to be more specific about the error here?

**Review by TomerYakir — CHANGES_REQUESTED** [review]
Added a few minor comments.
Do you suppose we can add an integration test to test out pinned server selection on mongonet?

File: wire.go (hunk 1 above)
Comment (tolsen) [reply wire.go:170]: I believe we would have errored when we constructed the MessageMessage in the first place.  That's why I treated this more like an internal contract violation.  I'll see how onerous it is to thread an error response all the way up

Comment (tolsen) [reply wire.go:170]: it's not that bad. i'll return an error

File: session.go (hunk 2 above)
Comment (tolsen) [reply session.go]: fixed!

**Comment by tolsen** [comment]
@TomerYakir comments addressed!

**Comment by tolsen** [comment]
> Added a few minor comments.
> Do you suppose we can add an integration test to test out pinned server selection on mongonet?

just saw this question now.  i'll look to see what can be easily added

**Comment by TomerYakir** [comment]
@louisaberger I added an int test for this and confirmed it's failing when the cursor isn't associated with the server address.
ready for review!

File: inttests/int_test_utils.go [hunk 3]
```diff
@@ -31,6 +32,7 @@ func insertDummyDocs(client *mongo.Client, numOfDocs int, ctx context.Context) e
 	if _, err := coll.InsertMany(ctx, docs); err != nil {
 		return fmt.Errorf("initial insert failed. err: %v", err)
 	}
+	fmt.Println("inserted", numOfDocs, "dummy docs")
```
Comment (louisaberger) [inline inttests/int_test_utils.go:35]: Do we still need this statement?

File: session.go [hunk 4]
```diff
@@ -125,6 +124,7 @@ func (s *Session) RespondToCommandMakeBSON(clientMessage Message, args ...interf
 	return s.RespondToCommand(clientMessage, doc2)
 }
 
+// do not call with OP_GET_MORE
```
Comment (louisaberger) [inline session.go:127]: Could you add a comment about _why_ we shouldn't call with OP_GET_MORE?

File: wire.go [hunk 5]
```diff
@@ -49,8 +53,15 @@ type ReplyMessage struct {
 	StartingFrom   int32
 	NumberReturned int32
 
-	Docs       []SimpleBSON
-	CommandDoc bsoncore.Document
+	Docs []SimpleBSON
+}
```
Comment (louisaberger) [inline wire.go:57]: Why do we now calculate this each time in the new function instead?

File: inttests/int_test_utils.go (hunk 3 above)
Comment (TomerYakir) [reply inttests/int_test_utils.go:35]: removed!

File: session.go (hunk 4 above)
Comment (tolsen) [reply session.go:127]: that function doesn't support OP_GET_MORE

File: wire.go (hunk 5 above)
Comment (tolsen) [reply wire.go:57]: it's a derived field.  I don't see how the below function CommandDoc() is even remotely expensive to call.

The idea is less state, fewer bugs.  This way I got to remove a bunch of code that was intended to keep this field up to date.  I found myself having to add more of that code.  It just seems really error prone.

**Review by louisaberger — APPROVED** [review]

**Pull request merged**